src/python/xml_parser-py.c

```diff
@@ -792,3 +790,247 @@ py_xml_parse_main_metadata_together(G_GNUC_UNUSED PyObject *self, PyObject *args
+            G_GNUC_UNUSED PyObject *args,
+            G_GNUC_UNUSED PyObject *kwds)
+{
+    _PkgIteratorObject *self = (_PkgIteratorObject *)type->tp_alloc(type, 0);
+    if (self) {
+        self->pkg_iterator = NULL;
+    }
+    return (PyObject *)self;
+}
+
+PyDoc_STRVAR(pkg_iterator_init__doc__,
+             "PkgIterator object\n\n"
+             ".. method:: __init__()\n\n"
+             "    Default constructor\n");
+
+static int
+pkg_iterator_init(_PkgIteratorObject *self, PyObject *args, PyObject *kwargs)
+{
+    char *primary_path;
+    char *filelists_path;
+    char *other_path;
+    PyObject *py_newpkgcb, *py_warningcb;
+    CbData cbdata;
+    GError *tmp_err = NULL;
+    static char *kwlist[] = {"primary", "filelists", "other", "newpkgcb",
+                             "warningcb", NULL};
+
+    if (!PyArg_ParseTupleAndKeywords(args, kwargs, "sssOO:pkg_iterator_init", kwlist,
+                                     &primary_path, &filelists_path, &other_path, &py_newpkgcb,
+                                     &py_warningcb)) {
+        return -1;
+    }
+
+    if (!PyCallable_Check(py_newpkgcb) && py_newpkgcb != Py_None) {
+        PyErr_SetString(PyExc_TypeError, "newpkgcb must be callable or None");
+        return -1;
+    }
+
+    if (!PyCallable_Check(py_warningcb) && py_warningcb != Py_None) {
+        PyErr_SetString(PyExc_TypeError, "warningcb must be callable or None");
+        return -1;
+    }
+
+    if (self->pkg_iterator) // reinitialization by __init__()
+        cr_PkgIterator_free(self->pkg_iterator, &tmp_err);
+    // TODO: correct thing to do with error?
+    if (tmp_err) {
+        nice_exception(&tmp_err, NULL);
+        return -1;
+    }
+
+    Py_XINCREF(py_newpkgcb);
+    Py_XINCREF(py_warningcb);
+    Py_XINCREF(cbdata.py_pkgs);
+
+    cr_XmlParserNewPkgCb ptr_c_newpkgcb = NULL;
+    cr_XmlParserWarningCb ptr_c_warningcb = NULL;
+
+    if (py_newpkgcb != Py_None)
+        ptr_c_newpkgcb = c_newpkgcb;
+    if (py_warningcb != Py_None)
+        ptr_c_warningcb = c_warningcb;
+
+    cbdata.py_newpkgcb = py_newpkgcb;
+    cbdata.py_pkgcb = NULL;  // TODO: What should go here?
+    cbdata.py_warningcb = py_warningcb;
+    cbdata.py_pkgs = PyDict_New();
+
+    self->pkg_iterator = cr_PkgIterator_new(
+        primary_path, filelists_path, other_path, ptr_c_newpkgcb, &cbdata, ptr_c_warningcb, &cbdata, &tmp_err);
+
+    // TODO: correct?
+    if (tmp_err) {
+        nice_exception(&tmp_err, NULL);
+        return -1;
+    }
+
+    if (self->pkg_iterator == NULL) {
+        PyErr_SetString(CrErr_Exception, "PkgIterator initialization failed");
+        return -1;
+    }
+
+    Py_XDECREF(py_newpkgcb);
+    Py_XDECREF(py_warningcb);
+    Py_XDECREF(cbdata.py_pkgs);
+
+    return 0;
+}
+
+static void
+pkg_iterator_dealloc(_PkgIteratorObject *self)
+{
+    GError *tmp_err;
+    if (self->pkg_iterator)
+        cr_PkgIterator_free(self->pkg_iterator, &tmp_err);
+    Py_TYPE(self)->tp_free(self);
```

Comment:
We also have to `free` the mallocated `cbdata`.
```suggestion
    if (self->cbdata) {
        Py_XDECREF(self->cbdata->py_newpkgcb);
        Py_XDECREF(self->cbdata->py_warningcb);
        Py_XDECREF(self->cbdata->py_pkgs);

        free(self->cbdata);
    }
    Py_TYPE(self)->tp_free(self);
```